working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+
+An instance variable in a subclass will always shadow a named parameter of the same name in a superclass constructor. That's why we do not forward to superclass constructors which have a required named parameter with such a name.
+
+Superclass constructors with optional positional parameters are hard to forward to. It would be wonderful if it was possible to have both optional positional parameters and named parameters on the same function.
+
+Existing mixin-application class forwarding constructors are just a special case of the general constructor forwarding functionality.
+
+We make constructors `const` only when the subclass declares no instance variables. We could allow final instance variables with no initializer expression (because then the constructor must initialize it) or with an initializer expression which is constant (because then it's valid in a `const` constructor class). The latter requirement is particularly fragile and useless. Changing the initializer expression of a variable from `final int x = a + 2;` to `final int x = a + b;` where `a` is a constant variable and  `b` is a non-constant variable with the value `2` will change the constructors of the class from constant to non-constant. Also, a final instance variable initialized to a constant value should probably just be a getter. We want to avoid accidentally making a class have constant constructors, because that may become a support burden for the author, so we only make constructors constant when the subclass is particularly simple, and where the current mixin application constructor forwarding makes the constructor constant. (It is not a *breaking* change to make more constructors constant in the future, but it is a *bad* change for library owners which intended a class to have non-constant constructors).
+
+You only get default constructors when you write *no* constructors. If you want to forward ten superclass constructors, and change the eleventh just a bit, or add an eleventh factory constructor, then you have to write all of the other ten in full. If this is too annoying, we may be able to add brief syntax for forwarding a constructor verbatim, say `Foo.bar = super.bar`.
+
+It's always possible to write exactly the same constructors manually, so this change does not introduce any new power. It's always possible to move away from default constructors to explicitly written constructors (except for classes used as mixins, which are always going to be highly restricted). As such, feature has no back-end impact, it can be implemented entirely in the front-end. Back-ends may want to tree-shake unused constructors, or unused constructors parameters, though.
```

Comment:
When you consider how this interacts with version constraints, it does give you a special power at the package level where a class inheriting a class from another package can then get a set of constructors determined by pub and not the class itself. Probably not a *good* power...

Reply:
True. That's always going to be the case. If a class adds an extra optional parameter to its unnamed constructor, then any forwarding constructor will also get that extra parameter. That's what forwarding means.

Reply:
Fair, though I think the scope of that is much smaller than having entirely new named constructors appear on a class.

Reply:
I have other issues with forwarding all named constructors, so I think I'll remove that again. Just the unnamed one, or nothing.

I should be easy to write `Foo.foo(super);` if you do want to forward anyway.